self_unenrollment_tool/lti1p3_utils.py

```diff
@@ -0,0 +1,55 @@
+from functools import wraps
+from logging import getLogger
+
+from django.http import HttpRequest, HttpResponseBadRequest
+from pylti1p3.contrib.django import (DjangoCacheDataStorage, DjangoDbToolConf,
+                                     DjangoMessageLaunch)
+
+logger = getLogger(__name__)
+
+
+class LtiResponseBadRequest(HttpResponseBadRequest):
+    def __repr__(self) -> str:
+        return f'No launch data associated with request'
+
+
+class CustomDjangoMessageLaunch(DjangoMessageLaunch):
+    # Override the default validate_deployment() method from DjangoMessageLaunch since we don't want
+    # to have to reconfigure the tool every time someone deploys it in a new course or sub-account.
+    def validate_deployment(self):
+        return self
+
+
+def get_message_launch(request: HttpRequest, launch_id: str=None):
+    tool_conf = DjangoDbToolConf()
+    launch_data_storage = DjangoCacheDataStorage()
+    if launch_id:
+        message_launch = CustomDjangoMessageLaunch.from_cache(launch_id, request, tool_conf,
+                                                            launch_data_storage=launch_data_storage)
+    else:
+        message_launch = CustomDjangoMessageLaunch(request, tool_conf, launch_data_storage=launch_data_storage)
+    return message_launch
+
+
+def get_launch_url(request: HttpRequest):
+    target_link_uri = request.POST.get('target_link_uri', request.GET.get('target_link_uri'))
+    if not target_link_uri:
+        raise Exception('Missing "target_link_uri" param')
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using `Exception` and `BaseException` can make your code prone to errors and difficult to maintain. Instead, we recommend using one of the [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html) or creating a custom exception class that is derived from `Exception` or one of its subclasses.